### DISCUSSION TITLE
What exactly is qLD?

### DISCUSSION THREAD

**Discussion by JeanElsner** [google-deepmind/mujoco#224] [Asking for Help] [answered]
According to the API documentation `mjData.qLD` is an _L'\*D\*L factorization of M_ whereas `mjData.qLDiagInv` is the inverse of the diagonal elements of the factorization. As the diagonal elements of the full matrix of qLD retrieved from `mj_fullM` are not ones as expected for an LDL decomposition I think qLD is actually L\*D. This seems to be true as qLD\*qlDiagInv results in a matrix with ones on the diagonal. This matrix however is also not a triangular matrix, which I think may be due to `mj_fullM`. In any case I am unable to reconstruct the dense representation of qM from these entities. I would expect  L'\*D\*L = qM where L = tril(qLD*qlDiagInv), D = diag(1/qlDiagInv) and the respective dense representations of qLD and qM.

I hope this is understandable, any help would be appreciated.

**Comment by yuvaltassa** [comment]
Both `qM` and `qLD` are the non-zero elements of branch-induced sparse matrices, as in Chapter 6.4 of [Featherstone (2008)](https://link.springer.com/book/10.1007/978-1-4899-7560-7).

[mj_fullM](https://mujoco.readthedocs.io/en/latest/APIreference.html#mj-fullm) reconstructs the dense, `nv x nv` representation. Here is the implementation, it will give you insight into the memory structure: 

```C
// convert sparse inertia matrix M into full matrix
void mj_fullM(const mjModel* m, mjtNum* dst, const mjtNum* M) {
  int i, j, adr = 0, nv = m->nv;
  mju_zero(dst, nv*nv);

  for (i=0; i<nv; i++) {
    j = i;
    while (j>=0) {
      dst[i*nv+j] = M[adr];
      dst[j*nv+i] = M[adr];
      j = m->dof_parentid[j];
      adr++;
    }
  }
}
```

**Comment by JeanElsner** [comment]
Yes, thank you. What I gather from this piece of code is that `mj_fullM` will always reconstruct a symmetric matrix. For an LDL decomposition the L matrix however is supposed to be a unit triangular matrix. But even if you make L triangular yourself the main diagonal entries are not 1. They become 1 if you multiply a diagonal matrix with qLDiagInv elements (hence I think qLD is L\*D). So far so good, but even then L'\*D\*L does not equal the dense representation of qM.

Maybe a little more concrete: how do qLD and qLDiagInv relate to qM (concerning dense nv x nv representations).
My interpretation is that the following should hold:
```
#!/usr/bin/env python3
import mujoco
import numpy as np

XML = """<mujoco>
  <worldbody>
    <body name="test" pos="0 0 .5" euler="45 45 0">
      <geom type="box" size=".1 .1 .1" rgba="1 1 1 .8" />
      <body pos="0 0 .5">
        <joint />
        <geom type="box" size=".1 .1 .1" rgba="1 1 1 .8" />
        <body pos="0 0 .5">
          <joint />
          <geom type="box" size=".1 .1 .1" rgba="1 1 1 .8" />
          <body pos="0 0 .5">
            <joint />
            <geom type="box" size=".1 .1 .1" rgba="1 1 1 .8" />
              <body pos="0 0 .5">
                <joint />
                <geom type="box" size=".1 .1 .1" rgba="1 1 1 .8" />
            </body>
          </body>
        </body>
      </body>
    </body>
  </worldbody>
</mujoco>"""

model = mujoco.MjModel.from_xml_string(XML)
data = mujoco.MjData(model)
mujoco.mj_step(model, data)

LD = np.zeros((model.nv, model.nv))
D = np.diag(1/data.qLDiagInv)
M = np.zeros((model.nv, model.nv))

# Get dense representation of qM and qLD
mujoco.mj_fullM(model, LD, data.qLD)
mujoco.mj_fullM(model, M, data.qM)

# Get unit triangular matrix from LD
L = np.tril(LD@np.diag(data.qLDiagInv))

# L'*T*L = M
print(np.allclose(L.T@D@L, M))
```

**Comment by yuvaltassa** [comment, marked as answer]
Yes, `LD` storage needs clarification. It is the lower triangle of the LDL factorisation with D replacing the unit diagonal. 
Using Python notation, if
```py
M = np.zeros((nv, nv))
LD = np.zeros((nv, nv))
mujoco.mj_fullM(model, M, data.qM)
mujoco.mj_fullM(model, LD, data.qLD)
``` 
Then for
```py
L = np.tril(LD, -1) + np.eye(nv)
D = np.diag(np.diag(LD))
```
we get
```py
L.T @ D @ L == M
```
to numerical precision. The `qLDiagInv` and `qLDiagSqrtInv` are there just to avoid taking roots and inverses during the solve, which is an important operation.

NB we have the upper triangle on the _left_ and the lower triangle on the _right_, since this is zero-fill-in _reverse Cholesky_, what Featherstone calls "LTL", rather than the LDL you'll find in standard textbooks.

**Reply by JeanElsner** [reply]
Ah, thank you so much, this clarifies it completely. A simple and compact scheme, but I didn't think of this.

**Comment by kevinzakka** [comment]
This is super informative, thanks @yuvaltassa !
